Source/WebCore/page/WorkerNavigator.cpp

```diff
@@ -62,11 +62,16 @@ GPU* WorkerNavigator::gpu()
     if (!m_gpuForWebGPU) {
         auto scriptExecutionContext = this->scriptExecutionContext();
         if (scriptExecutionContext->isWorkerGlobalScope()) {
+            RefPtr<WebGPU::GPU> gpu;
             WorkerGlobalScope& workerGlobalScope = downcast<WorkerGlobalScope>(*scriptExecutionContext);
-            if (!workerGlobalScope.graphicsClient())
-                return nullptr;
-
-            RefPtr gpu = workerGlobalScope.graphicsClient()->createGPUForWebGPU();
+            if (auto type = workerGlobalScope.type(); type == WorkerGlobalScope::Type::SharedWorker || type == WorkerGlobalScope::Type::ServiceWorker)
```

Comment:
The constructor for `SharedWorkerThreadProxy` creates a `WorkerClient` (implementing `GraphicsClient`), so I would expect the `else` path to 'just work' for that case.

I think we could also do the same in `ServiceWorkerThreadProxy` and hopefully not need this first path at all?

Reply:
It creates a nullptr `WorkerClient`:

```
(lldb) bt
(lldb) bt
frame #0: WebCore`WebCore::ChromeClient::createWorkerClient(this=0x000000010bb0bd50, (null)=0x000000010bb1ee60) at ChromeClient.h:297:98 [opt]
frame #1: WebCore`WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy(this=0x000000010bb278e0, page=<unavailable>, sharedWorkerIdentifier=WebCore::SharedWorkerIdentifier @ 0x000000016d6c9a40, clientOrigin=<unavailable>, workerFetchResult=<unavailable>, workerOptions=<unavailable>, initializationData=<unavailable>, cacheStorageProvider=<unavailable>) at SharedWorkerThreadProxy.cpp:116:46 [opt]
frame #2: WebKit`WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker(WebCore::ClientOrigin&&, WTF::ObjectIdentifierGeneric<WebCore::SharedWorkerIdentifierType, WTF::ObjectIdentifierMainThreadAccessTraits>, WebCore::WorkerOptions&&, WebCore::WorkerFetchResult&&, WebCore::WorkerInitializationData&&) [inlined] WTF::Ref<WebCore::SharedWorkerThreadProxy, WTF::RawPtrTraits<WebCore::SharedWorkerThreadProxy>, WTF::DefaultRefDerefTraits<WebCore::SharedWorkerThreadProxy>> WebCore::SharedWorkerThreadProxy::create<WTF::Ref<WebCore::Page, WTF::RawPtrTraits<WebCore::Page>, WTF::DefaultRefDerefTraits<WebCore::Page>>, WTF::ObjectIdentifierGeneric<WebCore::SharedWorkerIdentifierType, WTF::ObjectIdentifierMainThreadAccessTraits>&, WebCore::ClientOrigin&, WebCore::WorkerFetchResult, WebCore::WorkerOptions, WebCore::WorkerInitializationData, WebKit::WebCacheStorageProvider&>(args=0x000000016d6c9e28, args=<unavailable>, args=0x000000016d6ca150, args=0x000000016d6ca1a8, args=0x000000016d6ca198, args=0x000000016d6ca270, args=0x000000010b01c560) at SharedWorkerThreadProxy.h:53:114 [opt]
frame #3: WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker(this=0x000000010b8d8fe0, origin=0x000000016d6ca150, sharedWorkerIdentifier=WebCore::SharedWorkerIdentifier @ x24, workerOptions=0x000000016d6ca198, workerFetchResult=0x000000016d6ca1a8, initializationData=0x000000016d6ca270) at WebSharedWorkerContextManagerConnection.cpp:149:36 [opt]
```

ChromeClient::createWorkerClient is implemented as:
```
virtual std::unique_ptr<WorkerClient> createWorkerClient(SerialFunctionDispatcher&) { return nullptr; }
```

Reply:
Ah, I see!

I guess this really exposes an existing problem where the `createWorkerClient` function is tied into `ChromeClient`, where it probably should be a standalone `WorkerClientProvider`.

`WebSharedWorkerContextManagerConnection::launchSharedWorker` creates a page with an `EmptyChromeClient`, but does override a bunch of the other client/providers to be valid `WebKit` instances.

I think ideally we'd be able to do the same here, and have `pageConfiguration.workerClientProvider = makeUniqueRef<WorkerClientProvider>();`, and then the construction of the `WorkerClient` itself would work.

Reply:
Oh nice! I’ll try that instead

Reply:
@mattwoodrow I looked into implementing this and it seemed like WebWorkerClient didn't need the WebPage, just a WebCore::Page, so changing WebPage -> WebCore::Page allowed for Shared and Service workers to trivially construct a WebWorkerClient

Change is far simpler now and the conditional branch you mentioned above indeed just works now for shared and service workers :)